### PR TITLE
Add ViT NEPA

### PR DESCRIPTION
**Pull request by sbucaille**
# What does this PR do?
Fixes #42977

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@yonigozlan

### CODE REVIEW

**Comment by sbucaille** [comment]
Hey @yonigozlan that was quick, the PR is ready for a review. This consists mostly in wrapping ViT with the NEPA self supervised framework, pretty straightforward. Test suite is mostly the same as ViT, with an added test for the `ViTNepaForPreTraining` class.
@stevhliu let me know what you think about the docs, same, mostly pasted from ViT and adapted with a small description of the framework

**Comment by ArthurZucker** [comment]
cc @molbap as well if you have time to review!

File: docs/source/en/model_doc/vit_nepa.md [hunk 1]
```diff
@@ -0,0 +1,113 @@
+<!--Copyright 2025 The HuggingFace Team. All rights reserved.
```
Comment (molbap) [inline docs/source/en/model_doc/vit_nepa.md]: happy new year hehe 
```suggestion
<!--Copyright 2026 The HuggingFace Team. All rights reserved.
```

File: src/transformers/models/vit_nepa/modular_vit_nepa.py [hunk 2]
```diff
@@ -0,0 +1,609 @@
...
+    # Normalize the attention scores to probabilities.
+    attn_weights = nn.functional.softmax(attn_weights, dim=-1, dtype=torch.float32).to(query.dtype)
+
+    # This is actually dropping out entire tokens to attend to, which might
+    # seem a bit unusual, but is taken from the original Transformer paper.
+    attn_weights = nn.functional.dropout(attn_weights, p=dropout, training=module.training)
+
+    # Mask heads if we want to
+    if attention_mask is not None:
+        attn_weights = attn_weights * attention_mask
```
Comment (molbap) [inline src/transformers/models/vit_nepa/modular_vit_nepa.py]: sorry, came back to this. I'm a bit confused at the presence of both `causal_mask` and `attention_mask` here, especially since at this stage, attention mask has to be _added_ to the weights pre-softmax, right?

File: src/transformers/models/vit_nepa/modular_vit_nepa.py [hunk 3]
```diff
@@ -0,0 +1,609 @@
...
+    ) -> tuple[torch.Tensor, torch.Tensor]:
+        batch_size = hidden_states.shape[0]
+        new_shape = batch_size, -1, self.num_attention_heads, self.attention_head_size
+
+        key_layer = self.key(hidden_states).view(*new_shape).transpose(1, 2)
+        value_layer = self.value(hidden_states).view(*new_shape).transpose(1, 2)
+        query_layer = self.query(hidden_states).view(*new_shape).transpose(1, 2)
+
+        query_layer = self.q_norm(query_layer)
+        key_layer = self.k_norm(key_layer)
```
Comment (molbap) [inline src/transformers/models/vit_nepa/modular_vit_nepa.py]: this runs regardless of the `config.qk_norm` value, right? right now it exists and defaults to `False`, it should inform of the normalization rather (or not exist)

File: src/transformers/models/vit_nepa/modular_vit_nepa.py [hunk 4]
```diff
@@ -0,0 +1,609 @@
...
+            embeddings = embeddings * (1.0 - mask) + mask_tokens * mask
+
+        # add the [CLS] token
+        cls_tokens = self.cls_token.expand(batch_size, -1, -1)
+        embeddings = torch.cat((cls_tokens, embeddings), dim=1)
+        embeddings_clean = torch.cat((cls_tokens, embeddings_clean), dim=1)
+
+        embeddings = self.dropout(embeddings)
+
+        return embeddings, embeddings_clean
```
Comment (molbap) [inline src/transformers/models/vit_nepa/modular_vit_nepa.py]: this returns a tuple of Tensors (so need to change the hinting). Also returning tuples in module forwards is sometimes causing issues with tracing modules, if both are needed (for training) it would be worth having something like 

```python
     @dataclass
     class ViTNepaEmbeddingOutput:
         embeddings: torch.Tensor
         clean_embeddings: torch.Tensor
```

File: src/transformers/models/vit_nepa/modular_vit_nepa.py [hunk 5]
```diff
@@ -0,0 +1,609 @@
...
+    pass
+
+
+def eager_attention_forward(
+    module: nn.Module,
+    query: torch.Tensor,
+    key: torch.Tensor,
+    value: torch.Tensor,
+    attention_mask: Optional[torch.Tensor],
+    scaling: float,
```
Comment (molbap) [inline src/transformers/models/vit_nepa/modular_vit_nepa.py]: if scaling is None, should fallback (like in vit eager)

File: src/transformers/loss/loss_vit_nepa.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    hidden_states_in = hidden_states_in.detach()
+
+    if shift:
+        # shift one step forward
+        p = hidden_states_out[:, :-1, :]  # predict next
+        z = hidden_states_in[:, 1:, :]  # target is next hidden state
+    else:
+        # same-position matching
+        p = hidden_states_out
+        z = hidden_states_in
```
Comment (molbap) [inline src/transformers/loss/loss_vit_nepa.py]: do we need the same-position matching? I'm not 100% familiar with the NEPA training setup but we don't need MAE no? (feel free to correct me :laughing: )

**Review by molbap — COMMENTED** [review]
Hey @sbucaille ! Thanks for the addition, left a few questions to clarify, if you can help me understand a bit!

File: docs/source/en/model_doc/vit_nepa.md [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+</hfoption>
+</hfoptions>
+
+## Notes
+
+- The ViT NEPA models are pretrained using self-supervised learning on large-scale datasets without labels.
+- After pretraining, the models are fine-tuned on downstream tasks such as image classification.
+- The models utilize standard Vision Transformer architectures with modifications to support the NEPA training objective.
+- Use [`ViTImageProcessorFast`] to resize (or rescale) and normalize images to the expected size.
+- The patch and image resolution are reflected in the checkpoint name. For example, `SixAILab/nepa-base-patch14-224-sft`, is the **base-sized** architecture with a patch resolution of 16x16 and fine-tuning resolution of 224x224.
```
Comment (stevhliu) [inline docs/source/en/model_doc/vit_nepa.md]:
```suggestion
- The patch and image resolution are reflected in the checkpoint name. For example, [SixAILab/nepa-base-patch14-224-sft](https://huggingface.co/SixAILab/nepa-base-patch14-224-sft), is the **base-sized** architecture with a patch resolution of 16x16 and fine-tuning resolution of 224x224.
```

**Review by stevhliu — APPROVED** [review]
thanks, just one tiny nit!

File: docs/source/en/model_doc/vit_nepa.md (hunk 1 above)
Comment (sbucaille) [reply docs/source/en/model_doc/vit_nepa.md]: Happy new year ! Fixed licences and doc in b6eeb16ec5ba9a419fe39a6d7c3788acff23afa4

File: src/transformers/models/vit_nepa/modular_vit_nepa.py (hunk 2 above)
Comment (sbucaille) [reply src/transformers/models/vit_nepa/modular_vit_nepa.py]: I believe the `attention_mask` is the head mask which would be a leftover from pre-v5 since head mask support has been dropped if I'm correct. `causal_mask` in this case is the for the model to attend previous patchs to predict the next (as per the goal of the paper). Since head masks are dropped, `attention_mask` is always `None`. But you are right, this is from the original code and the position of these instructions should be above, I fixed it in 835d2046fc086fe3602a2f3153884d14487b867b let me know what you think

File: src/transformers/models/vit_nepa/modular_vit_nepa.py (hunk 3 above)
Comment (sbucaille) [reply src/transformers/models/vit_nepa/modular_vit_nepa.py]: Good catch, fixed in af6bc0fd2149e0cf2430db04c7c1279a95773380

File: src/transformers/models/vit_nepa/modular_vit_nepa.py (hunk 4 above)
Comment (sbucaille) [reply src/transformers/models/vit_nepa/modular_vit_nepa.py]: Was not aware of such tracing problems thanks for the heads up, I added the dataclass in 77e42685926fde07e02f8f839260c15a15db53d3

File: src/transformers/models/vit_nepa/modular_vit_nepa.py (hunk 5 above)
Comment (sbucaille) [reply src/transformers/models/vit_nepa/modular_vit_nepa.py]: Fixed in 5653d139236a9baaa79fcfcce1739793b2802c9e

File: src/transformers/loss/loss_vit_nepa.py (hunk 6 above)
Comment (sbucaille) [reply src/transformers/loss/loss_vit_nepa.py]: You are right, the same-position branch should not exist, shift is never different from True, I removed it in c3279f68f2205e57fa6ed94f1c77b8729aab5af6

File: docs/source/en/model_doc/vit_nepa.md (hunk 7 above)
Comment (sbucaille) [reply docs/source/en/model_doc/vit_nepa.md]: Fixed docs in b6eeb16ec5ba9a419fe39a6d7c3788acff23afa4 thanks !

**Comment by sbucaille** [comment]
Hey @molbap , thanks for your comments ! I addressed them, let me know if it looks good to you !

**Comment by sbucaille** [comment]
Hey @molbap , just bumping this PR as someone asked if it could be merged soon in the original issue
